# Исключение

*Throw*

![](<../../../.gitbook/assets/image (100) (1) (1) (1) (1) (1) (1) (1) (2) (87).png>)

![](<../../../.gitbook/assets/image (126).png>)

Элемент позволяет вызвать исключение в сценарии автоматизации. Вызов исключения может быть полезен в тестировании ошибочных сценариев и помогает настроить обработку исключений разных типов.


## Свойства

Описание общих свойств см. в разделе [Свойства элемента](https://docs.primo-rpa.ru/primo-rpa/primo-studio/process/elements#svoistva-elementa).

**Группа «Исключение»**:

> *Обязательно должно быть заполнено либо свойство «Сообщение», либо «Переменная исключения». Остальные свойства опциональные.*

1. **Сообщение\*** *[String]*: Сообщение создаваемого исключения. Пример: `"Создаем пользовательское исключение"`.
2. **Переменная исключения**\* *[[System.Exception](https://learn.microsoft.com/ru-ru/dotnet/fundamentals/runtime-libraries/system-exception)]*: Название переменной, хранящей исключение. Позволяет использовать для вызова уже существующее исключение. Если переменная указана, то свойства **Код**, **Сообщение** и **Тип исключения** заполнять не надо - значения будут взяты из переменной. Параметр добавлен в версии 1.23.11.
3. **Тип исключения** *[LTools.Common.Exceptions.ExceptionTypes]*: Тип исключения. По умолчанию `General` - общее исключение. Чтобы выбрать другой тип, нажмите на выпадающий список значений: 
   * ArgumentException - недопустимый аргумент, который передается методу.
   * NullReference - попытка обращения по нулевой ссылке.
   * IO Exception - ошибка ввода-вывода. Например, когда не найден указанный файл или папка.
   * Invalid Operation - 	вызов метода недопустим в текущем состоянии объекта.
   * Time Out - истекло время, выделенное для операции.
   * Custom - пользовательское исключение.
   * Business - исключение, связанное с бизнес-логикой. Например, когда определенные данные, которые необходимы в сценарии автоматизации, являются неполными или отсутствуют.
   * Application - исключение, которое вызвано технической проблемой, например, приложение не отвечает.
5. **Код** *[Int32]*: Код исключения. Код указывается пользователем на свое усмотрение, заданных кодов нет. Пример: `1`.


## Learning

На странице [Learning](https://github.com/PrimoRPA/Learning) доступен RPA-проект, демонстрирующий работу элемента совместно с Try-Catch.

1. Скачайте архив со всеми обучающими материалами по ссылке: [Скачать архив Learning](https://github.com/PrimoRPA/Learning/archive/refs/heads/master.zip).
2. Распакуйте архив и откройте в Студии проект **StudioActivities**.
3. Выберите процесс `StudioActivities/Ru/Управление/Try-Catch Исключение.ltw` для просмотра.


## Только код

Пример использования элемента в процессе с типом **Только код** (Pure code):

{% tabs %}
{% tab title="C#" %}
```csharp
throw new Exception("ThrowText");
//Для изменения типа исключения следует вместо Exception указывать тот класс, который подходит в данной ситуации. Например, ApplicationException.
throw new ApplicationException("text2");
```
{% endtab %}

{% tab title="Python" %}
```python
raise Exception("ThrowText")
```
{% endtab %}

{% tab title="JavaScript" %}
```javascript
throw "ThrowText";
```
{% endtab %}
{% endtabs %}

